src/control/fix.lean

```diff
@@ -0,0 +1,332 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+
+import data.nat.up
+import data.stream.basic
+import logic.basic
+import order.basic
+import order.bounded_lattice
+import order.omega_complete_partial_order
+import tactic.apply
+
+/-!
+# Fixed point
+
+This module defines a generic `fix` operator for defining recursive
+computations that are not necessarily well-founded or productive.
+An instance is defined for `roption`. The proof of its laws relies
+on omega complete partial orders (ωCPO).
+
+## Main definition
+
+ * class `has_fix`
+ * class `lawful_fix`
+ * `roption.fix`
+-/
+
+universes u v
+
+open_locale classical
+variables {α : Type*} {β : α → Type*}
+
+/-- `has_fix α` gives us a way to calculate the fixed point
+of function of type `α → α`. -/
+class has_fix (α : Type*) :=
+(fix : (α → α) → α)
+
+open omega_complete_partial_order
+
+section prio
+
+set_option default_priority 100  -- see Note [default priority]
+
+/-- Laws for fixed point operator -/
+class lawful_fix (α : Type*) [omega_complete_partial_order α] extends has_fix α :=
+(fix_eq : ∀ {f : α →ₘ α}, continuous f → has_fix.fix f = f (has_fix.fix f))
+
+lemma lawful_fix.fix_eq' {α} [omega_complete_partial_order α] [lawful_fix α]
+  {f : α → α} (hf : continuous' f) :
+  has_fix.fix f = f (has_fix.fix f) :=
+lawful_fix.fix_eq (continuous.to_bundled _ hf)
+
+end prio
+
+namespace roption
+
+open roption nat nat.up
+
+section basic
+
+variables (f : (Π a, roption $ β a) → (Π a, roption $ β a))
+
+/-- series of successive, finite approximation of the fixed point of `f` -/
+def fix.approx : stream $ Π a, roption $ β a
+| 0 := ⊥
+| (nat.succ i) := f (fix.approx i)
+
+/-- loop body for finding the fixed point of `f` -/
+def fix_aux {p : ℕ → Prop} (i : nat.up p) (g : Π j : nat.up p, i < j → Π a, roption $ β a) : Π a, roption $ β a :=
+f $ λ x : α,
+assert (¬p (i.val)) $ λ h : ¬ p (i.val),
+g (i.succ h) (nat.lt_succ_self _) x
+
+/-- fixed point of `f` -/
```

Comment:
More words please. This looks like the main definition in the file. Is this the least fixed point? If so, describe (briefly, in informal notation) the laws that it supports. (Yes, I know they are written below, if you know where to look, but the statements aren't particularly enlightening.)